kernel: provide Platform abstract interface for device drivers which are designed to works seamlessly on PCI bus, ACPI and DTS

najumon1980 opened the pull request:
Provide Platform abstract interface for device drivers which are designed to work on multiple bus topologies such as PCI, ACPI or just DTS base. Driver developers can use the new PLATFORM_DT_DEV_INST_DEFINE macro instead of DEVICE_DT_INST_DEFINE in case they have to design drivers which need to works seamlessly on PCI bus, ACPI and DTS base.  Currently  there is no support for MSI and multi vectors. Following are some of the key features/highlight:

- Abstract underlying bus/parent complexity and providing a generic interface to driver developers. Using the new interface, driver developers need not want to bother about underlying bus interface and associated complexity. 
- Based on their parent node type, the interface will automatically enumerate resource such as MMIO and IRQ resources and configure them which will be completely transparent to device drivers
- The interface will take care of memory map of MMIO address in case platform support MMU
- The new PLATFORM_DT_DEV_INST_DEFINE  interface take ISR as on of the argument and will take care the ISR registration based on underlying bus/parent node  type(PCI/ACPI or just DTS). 
- The new interface add very little overhead (almost NIL) 
 
This PR is depend on PR: https://github.com/zephyrproject-rtos/zephyr/pull/62694

najumon1980 committed:
arch: x86: add interface for encode irq flags

add interface for encode irq flags from acpica to arch specfic.
Currently enabled only for x86 archiecture.

Signed-off-by: Najumon B.A <najumon.ba@intel.com>

najumon1980 committed:
lib: acpi: add device resource enum support

add device resource enumaration support such as irq and mmio.

Signed-off-by: Najumon B.A <najumon.ba@intel.com>

najumon1980 committed:
kernel: devices: add platform abstract support

add abstract interface for devices drivers to enumerate and
configure resource such as mmio and interrupt. This will help
drivers which are designed to work on various bus such as pci,
acpi or can me dts base.

Signed-off-by: Najumon B.A <najumon.ba@intel.com>

najumon1980 committed:
board: x86: adl: add acpi resource enumeration for uart driver

add acpi resource enumeration for ns16550 uart driver on Intel
raptor lake platform

Signed-off-by: Najumon B.A <najumon.ba@intel.com>

najumon1980 committed:
lib: acpi: add resource enumeration shell command

add resource enumeration and acpi method shell commands such as
retrieve mmio, interrupt and gpio resources.

Signed-off-by: Najumon B.A <najumon.ba@intel.com>

najumon1980 committed:
drivers: uart: ns16550: update with new platform interface

update ns16550 uart driver with new platform interface for
enumerate and configure device resources. This help the drivers
works on multiple bus topology such as pci, acpi or dts base.

Signed-off-by: Najumon B.A <najumon.ba@intel.com>

najumon1980 committed:
boards: ns16550: update dts with io mapped flags

update board dts files with io_mapped flag where ns16550 register
base is io mapped devices

Signed-off-by: Najumon B.A <najumon.ba@intel.com>

tbursztyka commented on dts/x86/intel/raptor_lake.dtsi:
```diff
@@ -44,6 +45,25 @@
 		#address-cells = <1>;
 	};
 
+	acpi0: acpi0 {
```
ACPI is not a hw bus. This is going against DTS idea.

I think we already discussed that.

najumon1980 replied:
removed ACPI node. In order to distinguish whether a driver use ACPI interface or not, one can use DT_NODE_HAS_PROP(n, acpi_uid)). 

@galak could you please have a look and share your though ?

mbolivar-ampere commented on dts/bindings/acpi/acpi.yaml:
```diff
@@ -0,0 +1,20 @@
+
+  acpi-uid:
+    type: string
+    description: Provides OSPM with a logical device ID that does not change
```
https://docs.zephyrproject.org/latest/build/dts/bindings-upstream.html#descriptions

najumon1980 replied:
updated

mbolivar-ampere commented on include/zephyr/dt-bindings/acpi/acpi.h:
```diff
@@ -0,0 +1,8 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define ACPI_IRQ_DETECT 0xFFFFFFFU
```
Needs include guard

najumon1980 replied:
updated

mbolivar-ampere commented on include/zephyr/acpi/acpi.h:
```diff
@@ -88,6 +88,7 @@ struct acpi_mmio_resource {
  */
 #define ACPI_DT_HID(node_id) DT_PROP_OR(node_id, hardware_id, "INVALD")
 
+
```
Something is wrong with this commit. It isn't doing anything related to the commit description.

najumon1980 replied:
this is due to dependency PR https://github.com/zephyrproject-rtos/zephyr/pull/62694 and rebased it

mbolivar-ampere reviewed (changes requested):
I don't know ACPI and don't have any general objections. Please fix a few minor things.

najumon1980 committed:
drivers: pci: update prt retrieve based on pnp id

update prt retrieve based on acpi pnp id instead of acpi device
path/name

Signed-off-by: Najumon B.A <najumon.ba@intel.com>

najumon1980 committed:
boards: x86: add pci controller node with acpi pnp id

add acpi pnp/hw id for pcie node to enable support for retreive
interrupt routing information for pci legacy interrupt via acpi

Signed-off-by: Najumon B.A <najumon.ba@intel.com>

najumon1980 committed:
drivers: rtc: add acpi based resource enumeration

add acpi based resource enumeration for retrieve rtc mmio and
interrupt resources

Signed-off-by: Najumon B.A <najumon.ba@intel.com>

najumon1980 committed:
board: x86:  add new qemu board with acpi devices

add a new qemu board with acpi devices such as  uart and rtc

Signed-off-by: Najumon B.A <najumon.ba@intel.com>

najumon1980 committed:
arch: x86: add a sperate timer interface for acpica lib

add a separate timer interface for acpica lib instead of using
system timer which might use driver interface such as HPET and this
might cause init priority issue if a driver which need to init
before system timer driver instantiate.

Signed-off-by: Najumon B.A <najumon.ba@intel.com>

najumon1980 committed:
build: upate west point to latest acpica commit

update west to latest acpica commit for timer interface changes

Signed-off-by: Najumon B.A <najumon.ba@intel.com>

najumon1980 committed:
arch: x86: disable early console if uart use acpi

disable early console if uart use acpi based resource enumeration

Signed-off-by: Najumon B.A <najumon.ba@intel.com>

najumon1980 committed:
tests: lib: acpi: update test cases for dev and resource enum

update test cases for device enumeration, mmio resource and irq
resources acpi interfaces.

Signed-off-by: Najumon B.A <najumon.ba@intel.com>

najumon1980 committed:
drivers: ns16550: remove parent init level dependency

remove parent init level dependency such as PRE_KERNEL or
POST_KERNEL. Uart driver init level change always to PRE_KERNEL

Signed-off-by: Najumon B.A <najumon.ba@intel.com>

finikorg commented on boards/x86/qemu_x86/qemu_x86_64_acpi.dts:
```diff
@@ -0,0 +1,263 @@
+		acpi-uid = "0";
+		ranges;
+
+	smbus0: smbus0 {
```
wrong indentation for smbus

finikorg commented on include/zephyr/arch/x86/x86_acpi_osal.h:
```diff
@@ -26,4 +26,9 @@ static inline void *acpi_rsdp_get(void)
 	return bios_acpi_rsdp_get();
 }
 #endif /* CONFIG_X86_EFI */
+
```
The commit has syntax error `sperate` -> `separate`

finikorg commented on lib/acpi/acpi.c:
```diff
@@ -217,6 +217,9 @@ static ACPI_STATUS dev_resource_enum_callback(ACPI_HANDLE obj_handle, UINT32 lev
 	}
 
 	child_dev = (struct acpi_dev *)&acpi.child_dev[acpi.num_dev++];
+	if (!dev_info->HardwareId.String) {
+		goto exit;
+	}
```
This check should be put after function which get device_info.

finikorg commented on lib/acpi/acpi_shell.c:
```diff
@@ -220,19 +226,93 @@ static int dump_prt(const struct shell *sh, size_t argc, char **argv)
+		shell_print(sh, "Device Path: %s", dev->path);
+
+		if (!acpi_device_mmio_get(dev, &mmio_res)) {
+
```
No need for empty line here

finikorg commented on lib/acpi/acpi_shell.c:
```diff
@@ -220,19 +226,93 @@ static int dump_prt(const struct shell *sh, size_t argc, char **argv)
+		}
+
+		if (!acpi_device_irq_get(dev, &irq_res)) {
+
```
No need for empty line here

finikorg commented on drivers/rtc/rtc_mc146818.c:
```diff
@@ -15,9 +15,10 @@
-#define RTC_STD_INDEX (DT_INST_REG_ADDR_BY_IDX(0, 0))
-#define RTC_STD_TARGET (DT_INST_REG_ADDR_BY_IDX(0, 1))
+#define RTC_STD_INDEX(dev)  PLATFORM_MMIO_GET(dev)
+#define RTC_STD_TARGET(dev) (PLATFORM_MMIO_GET(dev) + 1)
```
Is there better way?

finikorg commented on boards/x86/qemu_x86/qemu_x86_64_acpi.yaml:
```diff
@@ -0,0 +1,20 @@
+  - xtools
+simulation: qemu
+supported:
+  - acpi
```
The board `qemu_x86_64` does also support ACPI, what is the difference then?
